### PR TITLE
[RLlib] Issue 39791: PPO's new stack: explore=False not working.

### LINKED ISSUE #39791: [RLlib] PPO evaluation does not work when explore = False
### What happened + What you expected to happen

If I set explore =  False in the evaluation_config of PPO, the evaluation will fail due to: `KeyError: 'vf_preds'`, when set to True it works fine.

Screenshot from 2023-09-21 20-13-44

### Versions / Dependencies

Python Version: 3.10
Ray Version: 2.7.0
OS Version: Pop!_OS 22.04 LTS

### Reproduction script

```python


from ray.rllib.algorithms.ppo import PPOConfig

algo = (
    PPOConfig()
    .framework("torch")
    .evaluation(evaluation_interval=1, evaluation_duration=5, evaluation_config= {"explore" : False})
    .rollouts(num_rollout_workers=1)
    .resources(num_gpus=1)
    .environment(env="CartPole-v1")
    .build()
)

for i in range(20):
    result = algo.train()

```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by sven1977**
Issue 39791: PPO's new stack: explore=False not working.

Closes #39791 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/ppo/tf/ppo_tf_rl_module.py [hunk 1]
```diff
@@ -45,13 +35,17 @@ def _forward_exploration(self, batch: NestedDict) -> Mapping[str, Any]:
         if STATE_OUT in encoder_outs:
             output[STATE_OUT] = encoder_outs[STATE_OUT]
 
-        # Value head
+        # Value head.
+        # TODO (sven): Once we don't use Policy anymore for PPO (on new EnvRunner API),
```
Comment (kouroshHakha) [inline rllib/algorithms/ppo/tf/ppo_tf_rl_module.py:39]: For this exact reason we should not merge compute vf in forward inference and we should keep that separate from forward explore. Why are we merging the two?

Comment (sven1977) [reply rllib/algorithms/ppo/tf/ppo_tf_rl_module.py:39]: B/c right now, the new stack is NOT usable for anyone configuring `explore=False` in their config. Which means it's currently breaking a lot of use cases, where users do greedy inference on a PPO module.

Agreed that this is only a temporary fix, but we need to quickly make this work.

Once PPO moves to the EnvRunner API, we will move vf computations completely outside of all forward passes and handle it where it should be handled (on the learner or algorithm end).

Comment (kouroshHakha) [reply rllib/algorithms/ppo/tf/ppo_tf_rl_module.py:39]: I don't think the temporary fix should be here. There should be a place outside of this where conditioned on `explore=True` or `explore=False` it will call `forward_exploration` or `forward_inference`. This is a hacky fix which goes against the design principles of RLModules.

Comment (sven1977) [reply rllib/algorithms/ppo/tf/ppo_tf_rl_module.py:39]: So why did this violation against the design principles of RLModule make it into the library in the first place?

In other words: Why were `forward_exploration()` and `forward_inference()` already (before this PR) behaving so differently and return either `vf_preds` or not returning `vf_preds`? How would an unsuspecting user have known that only `forward_exploration()` returns `vf_preds`, but - for some magic reason - `forward_inference()` does not?
Bottom line: The root cause of this problem is being addressed right now by one of Simon's PR's, in which we make PPO independent of any old Policy class and can then perform the vf computations completely independently of the `forward_inference/exploration` passes.

Comment (kouroshHakha) [reply rllib/algorithms/ppo/tf/ppo_tf_rl_module.py:39]: `forward_inference` is for running inference (when the model training is done) so it doesn't need the vf; you just need actions. `forward_exploration` is the forward method you use during sampling but during training. That's the mental model. 

Regarding the bottom-line, shouldn't we wait for that PR to land and fix the root cause instead of this fix?

Comment (sven1977) [reply rllib/algorithms/ppo/tf/ppo_tf_rl_module.py:39]: I disagree about the mental model, though. Both `forward_inference()` and `forward_exploration()` should be usable interchangeably within a sampling procedure supposed to collect data for training.

This is besides your point about fixing the root cause, which I 100% agree with, but it's a larger process that may take another couple of days/week(s) to be fully functional and tested. So in the meantime, can we just merge this fix and allow PPO to run normally again? It's currently broken for any user that wants to set `explore=False` in the eval workers, which is a completely sensible thing to do! This is not a rare usecase and therefore I made this issue P1:
https://github.com/ray-project/ray/issues/39791
